pulsar-transaction/buffer/src/main/java/org/apache/pulsar/transaction/buffer/impl/PersistentTransactionBuffer.java

```diff
@@ -0,0 +1,308 @@
+            } catch (ManagedLedgerException | TransactionNotSealedException e) {
+                readerFuture.completeExceptionally(e);
+            }
+            return null;
+        }).exceptionally(e -> {
+            readerFuture.completeExceptionally(e);
+            return null;
+        });
+
+        return readerFuture;
+    }
+
+    @Override
+    public CompletableFuture<Void> commitTxn(TxnID txnID, long committedAtLedgerId, long committedAtEntryId) {
+        CompletableFuture<Void> commitFuture = new CompletableFuture<>();
+
+        Marker commitMarker = Marker.builder().txnID(txnID).status(TxnStatus.COMMITTED).build();
+        ByteBuf marker = Unpooled.wrappedBuffer(commitMarker.serialize());
+
+        txnCursor.getTxnMeta(txnID, false).thenCompose(meta -> {
+            publishMessage(marker, meta.lastSequenceId() + 1).thenCompose(position -> {
+                txnCursor.commitTxn(committedAtLedgerId, committedAtEntryId, txnID, position).thenCompose(v -> {
+                    commitFuture.complete(null);
+                    return null;
+                }).exceptionally(e -> {
+                    commitFuture.completeExceptionally(e);
+                    return null;
+                });
+                return null;
+            }).exceptionally(e -> {
+                commitFuture.completeExceptionally(e);
+                return null;
+            });
+            return null;
+        }).exceptionally(e -> {
+            commitFuture.completeExceptionally(e);
+            return null;
+        });
+
+        return commitFuture;
+    }
+
+    @Override
+    public CompletableFuture<Void> abortTxn(TxnID txnID) {
+        CompletableFuture<Void> abortFuture = new CompletableFuture<>();
+
+        Marker abortMarker = Marker.builder().txnID(txnID).status(TxnStatus.ABORTED).build();
+        ByteBuf marker = Unpooled.wrappedBuffer(abortMarker.serialize());
+
+        txnCursor.getTxnMeta(txnID, false).thenCompose(meta -> {
+            publishMessage(marker, meta.lastSequenceId() + 1).thenCompose(position -> {
+                txnCursor.abortTxn(txnID).thenCompose(v -> {
+                    abortFuture.complete(null);
+                    return null;
+                }).exceptionally(e -> {
+                    abortFuture.completeExceptionally(e);
+                    return null;
+                });
+                return null;
+            }).exceptionally(e -> {
+                abortFuture.completeExceptionally(e);
+                return null;
+            });
+            return null;
+        }).exceptionally(e -> {
+            abortFuture.completeExceptionally(e);
+            return null;
+        });
+
+        return abortFuture;
+    }
+
+    private CompletableFuture<Position> publishMessage(ByteBuf msg, long sequenceId) {
+        CompletableFuture<Position> publishFuture = new CompletableFuture<>();
+
+        CountDownLatch latch = new CountDownLatch(1);
```

Comment:
can you explain why do you need a latch here?

Reply:
The publish operation is run in a thread and I need to wait for the publish operation done to ensure every operation succeeds or failed.